### ISSUE TITLE
Multiple monitors results in very large screen sizes

### ISSUE THREAD

**Issue by Mailaender** [adventuregamestudio/ags#258]
On 3.4.0.3 (Alpha) the game detects very large (combined) screen sizes when multiple monitors are used. I tried tweaking `acsetup.cfg` but my settings (like no full screen, letterbox yes/no) seem to get ignored. Running on openSUSE 13.2 (Linux).

```
AGS: Resonance
AGS: Checking for disk space
AGS: Initializing resolution settings
AGS: Device display resolution: 3286 x 1080
AGS: Game native resolution: 320 x 240 (32 bit) letterbox-by-design
AGS: Game settings: windowed = no, screen def: max, screen size: 0 x 0, match device ratio: yes, frame placement: center
AGS: Init gfx driver
AGS: Created graphics driver: Allegro/DX5
AGS: Supported gfx modes (32-bit): 
        640x480;800x600;800x600;1024x768;1366x768;
AGS: Supported gfx modes (24-bit): 
        640x480;800x600;800x600;1024x768;1366x768;
AGS: Requested gfx filter: StdScale, filter scaling: max uniform
AGS: Applying scaling filter: StdScale
AGS: Switching to graphics mode
AGS: Attempting to find nearest supported resolution for screen size 3286 x 1080 (32-bit) fullscreen
AGS: Could not find compatible graphics mode
AGS: Attempting to find nearest supported resolution for screen size 3286 x 1080 (24-bit) fullscreen
AGS: Could not find compatible graphics mode
AGS: Attempting to find nearest supported resolution for screen size 3286 x 1048 (32-bit) windowed
AGS: Attempt to switch gfx mode to 3286 x 1048 (32-bit) windowed; game frame: 1280 x 960, frame placement: center
AGS: Succeeded. Using gfx mode 3286 x 1048 (32-bit) windowed
        filter dest (1003, 44, 2282, 1003 : 1280 x 960), render dest (1003, 44, 2282, 1003 : 1280 x 960)
AGS: Preparing graphics mode screen
AGS: Initializing colour conversion
AGS: Check for preload image
AGS: Initialize sprites
AGS: Set up screen
AGS: Initialize game settings
AGS: Prepare to start game
```

This doesn't happen on stable version 3.3.4.1.

**Comment by ivan-mogilko** [comment]
This is a bug; AGS does not detect multiple monitors.

Regarding config: you probably edit wrong parameters, I changed the way they are stored in config file for 3.4.0. An oversight on my part is that readme in this repository still mentions old ways.

Please refer to to this pull request explanation, it lists all new parameters:
https://github.com/adventuregamestudio/ags/pull/190

As a temporary workaround for your problem, you may try using following config:
<pre>
[graphics]
windowed = 0
screen_def = scaling
filter = stdscale
filter_scaling = N ; where N is an integer scaling factor that suits you
</pre>

**Comment by 256colour** [comment]
I've ran into a similar bug to this several times now in the Linux OpenGL port we've been discussing. If another monitor is connected in "extend" mode (e.g. extend the desktop), launching a game in fullscreen produces strange windowing behaviour. On my machine it creates a small un-scaled window on the main desktop. This is on v.3.4.1.15 of the interpreter btw.

The workaround, so far, has been to disable one of the monitors before launching the game in fullscreen. Is there a better way of dealing with this problem that I'm missing?

**Comment by 256colour** [comment]
Responding to my own post - this appears to be a limitation of Allegro 4.x, as it cannot detect the number of attached monitors.

**Comment by 256colour** [comment]
Updating this thread on behaviour I'm seeing:
- On my machine with a 2560x1440 primary display and 1280x1024 secondary display, the Software Renderer produces a fullscreen mode on the *secondary monitor* (unexpected behaviour) at the *correct* resolution.
- The OGL renderer (using Toojays' fork) produces a fullscreen mode on the *primary monitor* (expected behaviour) at the *incorrect* resolution of 1280x1024.

**Comment by ivan-mogilko** [comment]
> does someone have a suggestion on how I can rewrite the Linux/OGL resolution detection scheme so that it produces correct window sizes on dual-monitor systems?

There's a virtual method added for getting system display resolutions, which is used with OpenGL renderer called AGSPlatformDriver::GetSystemDisplayModes.

**Comment by 256colour** [comment]
Just thought I'd keep this thread updated with how things are progressing with multi-monitor on Linux. I'm struggling to understand how [@toojays opengl branch](https://github.com/adventuregamestudio/ags/issues/510) hooks in with xwindows to detect and set resolutions.. but I've found some codefrags [here](http://www.cplusplus.com/forum/unices/108795/) and [here](https://tronche.com/gui/x/xlib/display/opening.html) that suggest a method for getting the display properties through xwindows.

I'm absolutely open to ideas/suggestions on how to implement this from anyone.

**Comment by 256colour** [comment]
Update:
I'm throwing in the towel on trying to code this feature. Using "xrandr" to create a virtual desktop, then splitting it two might have yielded some positive results... but I just don't have the AGS/C++ chops to pull it off.

**Comment by ivan-mogilko** [comment]
@256colour 
>  but I just don't have the AGS/C++ chops to pull it off.

could you tell what did you try and what were the problems you've met?

**Comment by ivan-mogilko** [comment]
@256colour, there's something strange in your code, this might be a mistake or some formatting went wring when copying, could you check that out?

<pre>Display* disp = XOpenDisplay(NULL:0.1);</pre>

The documentation of XOpenDisplay tells that the parameter is "char *displayname", should not it be taken in double quotes?

**Comment by 256colour** [comment]
Ack. Sorry, pasted some old buggy code. That should have read:
`Display* disp = XOpenDisplay(NULL);`

**Comment by Kirtai** [comment]
I've been running into this problem too, with fullscreen spanning all my monitors.
I have noticed however, that the linux AGS engine supplied with Primordia on GOG *does* get it right and properly displays fullscreen on my primary monitor.

Attached are logs from both the self compiled latest version and the version that comes with Primordia.

[logs.zip](https://github.com/adventuregamestudio/ags/files/4707013/logs.zip)

**Comment by Kirtai** [comment]
I got it to display properly by adding both of the 
```
screen_def=scaling
match_device_ratio
```
options to the config file under [graphics], however, the mouse pointer moves very slowly and jumpily.

**Comment by ivan-mogilko** [comment]
@Kirtai , since this ticket was first open AGS was added an OpenGL renderer support on Linux. According to your log your self-compiled engine started OpenGL, and unless Primordia has a custom engine it probably runs only in software mode. Could you tell, does this problem occur in software mode or only with OpenGL?

**Comment by ivan-mogilko** [comment]
There was an attempt to resolve this by @256colour as noted above: https://github.com/adventuregamestudio/ags/issues/258#issuecomment-454977854

But it was not finished, and looks like we forgot to include this ticket in a 3.5.0 milestone and did not try more.

**Comment by morganwillcock** [comment]
> Also, the mouse pointer moves very slowly and jumpily.

If you switch an existing game to use a newer engine (which would use OpenGL by default) or tell it to use OpenGL, the cursor control options may be the cause of this. Try adding:

<pre>
[mouse]
control_when=never
</pre>

**Comment by ericoporto** [comment]
I Imagine the order of things is (on Linux):
- [figure out which display the window is](https://github.com/spurious/SDL-mirror/blob/b7252b8c26cb64a563d6091bee4841e668d84a62/src/video/SDL_video.c#L1050)
- [get resolution list for the display](https://github.com/spurious/SDL-mirror/blob/b7252b8c26cb64a563d6091bee4841e668d84a62/src/video/x11/SDL_x11modes.c#L264)
- in case of success, pick biggest resolution, in any error, fallback to allegro way.

I don't know what the first step means when a game starts already in fullscreen (decision on the monitor to start the game).

**Comment by ivan-mogilko** [comment]
Game does not start in fullscreen, a window is created always before setting final mode.

**Comment by ericoporto** [comment]
From reading Allegro 4 source code it's correct [IF there are no Xinerama extensions](https://nouveau.freedesktop.org/wiki/MultiMonitorDesktop/).

Probably, in [`src/x/xsystem.c`](https://github.com/adventuregamestudio/lib-allegro/blob/allegro-4.4.3.1-agspatch/src/x/xsystem.c), under lib-allegro, it should be something like

```C
static int _xwin_sysdrv_get_desktop_resolution(int *width, int *height)
{
    XLOCK();
#if HAS_XINERAMA
    int screencount;
    XineramaScreenInfo *xinerama  = XineramaQueryScreens(_xwin.display, &screencount);
    if (xinerama) {
        *width = xinerama[_xwin.screen].width;
        *height = xinerama[_xwin.screen].height;
    } else {
#endif
        *width =  DisplayWidth(_xwin.display, _xwin.screen);
        *height = DisplayHeight(_xwin.display, _xwin.screen);
#if HAS_XINERAMA
    }
#endif
    XUNLOCK();
    return 0;
}
```
Testing this requires adding [Xinerama](https://github.com/liballeg/allegro5/blob/326fb49a6b979e43aa434f42925093d9b6e9b83d/CMakeLists.txt#L764) to the [CMakeLists.txt](https://github.com/adventuregamestudio/lib-allegro/blob/allegro-4.4.3.1-agspatch/CMakeLists.txt#L522). In the new allegro5 [many fullscreen functions](https://github.com/liballeg/allegro5/blob/master/src/x/xfullscreen.c) have xinerama and non-xinerama versions.

The situation appears to be very complex on the allegro5 source but I couldn't find a new `get_desktop_resolution` function, so maybe they removed.

**Comment by ivan-mogilko** [comment]
> The situation appears to be very complex on the allegro5 source but I couldn't find a new get_desktop_resolution function, so maybe they removed.

The Allegro5 API is vastly different.

**Comment by ericoporto** [comment]
Made a fix that seems to work alright on my end here:

https://github.com/adventuregamestudio/lib-allegro/compare/allegro-4.4.3.1-agspatch...ericoporto:xinerama-xsystem

We rewrite the lib-allegro `CMakeListst.txt` with our own! So to test one must download this [**`CMakeLists.txt ↓`**](https://gist.githubusercontent.com/ericoporto/aacbc7002ec1bd4e8c41943bf7cb5638/raw/98741748847817219c404d464d8ea8072a7a62f1/CMakeLists.txt), because I had to add Xinerama there, and place it in `Common/libsrc/allegro4`.

If you want directly test from this branch above, you can get this [**`FetchAllegro.cmake ↓`**](
https://gist.githubusercontent.com/ericoporto/6109075f59dcb589a0cb913cd7c40120/raw/5180f91437c7341ac519c249a18d6710196ac5d4/FetchAllegro.cmake), and place it in `CMake/FetchAllegro.cmake`.

This is all kinda hackish. 

---

From reading Allegro 4 original source code, it's correct [IF there are no Xinerama extensions](https://nouveau.freedesktop.org/wiki/MultiMonitorDesktop/).

In the new allegro5 [many fullscreen functions](https://github.com/liballeg/allegro5/blob/master/src/x/xfullscreen.c) have xinerama and non-xinerama versions. The situation appears to be very complex on the allegro5 source but I couldn't find a new `get_desktop_resolution` function, so maybe they removed.

Allegro 5 has a SDL platform in itself, which is kinda smart.

**Comment by ivan-mogilko** [comment]
@ericoporto when you say "add Xinerama", do you mean we must supply another header or library along?

**Comment by ericoporto** [comment]
it's a library to interface with an extension of X11 that is also called Xinerama [apparently](https://linux.die.net/man/3/xinerama). In Makefiles it could be `-lXinerama` but allegro is built with CMake.

It may be better to replace the functions we use from allegro to ask this information for SDL functions since they solve a lot of these complications. Maybe having SDL there already would help to work towards reducing Allegro dependency. Other possibility is adding a SDL platform to Allegro 4.4 itself like Allegro 5 has. This may be faster.

**Comment by ivan-mogilko** [comment]
When you say "this is hackish", how hackish is it?

Regarding SDL, I don't think it's wise to add SDL dependency just for this. If you're patching Allegro4 you could as well copy corresponding SDL code, since that's same C code after all.
Imo it's best to port to SDL properly.

**Comment by ericoporto** [comment]
Hackish means I am not sure what is the proper way to get this information. 

The code for the same thing in SDL is a ton more complex which makes me believe this problem is not simple.

**Comment by ivan-mogilko** [comment]
Well, since it looks like none of us know the exact solution, my thinking is it's better to
1) if there's a quick fix which works at least in general situation then patch allegro;
2) otherwise avoid taking on this in current version, and in terms of backend focus on properly moving to SDL. Spending time patching leaks, so to say, may take forever.

BTW, could be it's better to add this code not in Allegro but in AGS engine directly (Linux's platform driver)? Unless allegro must use this information too for some internal reasons.

**Comment by ericoporto** [comment]
For the system modes in ags is easy, the Linux platform driver already has a function to return modes (`AGSLinux::GetSystemDisplayModes`), so I could modify it.

For the resolution of the monitor I need to modify ags in some way since it uses `Engine/main/graphics_mode.cpp` function `get_desktop_size` directly, which in turn calls Allegro which will get the appropriate platform. I am not sure the best approach here.

**Comment by ivan-mogilko** [comment]
You could similarily call a platform driver, and default implementation would call allegro's get_desktop_size().

**Comment by ivan-mogilko** [comment]
> I have looked the ags code and it's still not clear for me if there's a limited amount of files that call Allegro or if it can be considered distributed on all of them.

It's limited to particular things. Most notable is use of bitmaps (which is mostly incapsulated in Bitmap class) and user input.

**Comment by ericoporto** [comment]
Cool, I found the `AGSPlatformDriver`, I can add one method there:
```C++
virtual Size GetDesktopSize();
```

and then do as you said and see how it works.

**Comment by ericoporto** [comment]
Sent a PR as #1095 , it's probably best let it sit there a little bit so some other Linux user/dev can take a look. 

I think maybe it's useful to also check and prefer a handling method using `XRandR` if available, since it's the newer and more updated supported method. It would fallback to xinerama if not available, and fallback to Xlib through allegro if not available. I could not yet figure out the code for this.

**Comment by ericoporto** [comment]
![Screenshot from 2021-05-03 17-52-03](https://user-images.githubusercontent.com/2244442/116932307-75d56f00-ac38-11eb-805d-d2a1db20a882.png)
![Screenshot from 2021-05-03 17-52-28](https://user-images.githubusercontent.com/2244442/116932315-7837c900-ac38-11eb-9d9e-687364747ed1.png)

a note that this is fixed in the master branch now that SDL2 has been merged, you can even alt+enter, drag the monitor to other monitor and alt+enter again to fullscreen in the other monitor and it works alright (I'm on Ubuntu 20.04 with GNOME!)

**Comment by ivan-mogilko** [comment]
@ericoporto thank you for telling, I don't have a dual monitor workplace at the moment.

There was another idea to let user configure the Display to place window on, but that's a separate issue. So I am closing this one.

Fixed in 3.6.0 (alpha version)
